assets/queries/cloudFormation/alarme_and_log_not_defined_for_root/query.rego

```diff
@@ -0,0 +1,74 @@
+package Cx
+
+CxPolicy [ result ] {
+
+	resourceMetricFilter := input.document[i].Resources
+    some name
+    resourceMetricFilter[name].Type == "AWS::Logs::MetricFilter"
+    
+    resourceAlarm := input.document[i].Resources
+    some nameAlarm
+    resourceAlarm[nameAlarm].Type == "AWS::CloudWatch::Alarm"
+   
+    checkAlarm(resourceMetricFilter[name].Properties.MetricTransformations[index].MetricNamespace, resourceAlarm[nameAlarm].Properties)
+
+    result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	     sprintf("Resources.%s.Properties.MetricTransformations.MetricNamespace", [name]),
+                "issueType":		"IncorrectValue",
+                "keyExpectedValue":  sprintf("'Resources.%s.Properties.MetricTransformations[%d].MetricNamespace' should have a valid Namespace in 'AWS::CloudWatch::Alarm')", [name, index]),
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern.MetricTransformations[%d].MetricNamespace' doesn't have a valid Namespace in 'AWS::CloudWatch::Alarm')", [name, index]),
+            }
+}
+CxPolicy [ result ] {
+
+	resourceMetricFilter := input.document[i].Resources
+   
+    resourceAlarm := input.document[i].Resources
+    some nameAlarm
+    resourceAlarm[nameAlarm].Type == "AWS::CloudWatch::Alarm"
+   
+    resourceSNS := input.document[i].Resources
+    some nameSNS
+    resourceSNS[nameSNS].Type == "AWS::SNS::Topic"
+   
+    checkTopic(resourceAlarm[nameAlarm].Properties.AlarmActions[index], nameSNS)
+
+    result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	     sprintf("Resources.%s.Properties.AlarmActions.Ref", [nameAlarm]),
+                "issueType":		"IncorrectValue",
+                "keyExpectedValue":  sprintf("'Resources.%s.Properties.AlarmActions[%d].Ref' should have a valid resource with 'AWS::SNS::Topic' type)", [nameAlarm, index]),
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern.AlarmActions[%d].Ref' doesn't have a valid resource with 'AWS::SNS::Topic' type)", [nameAlarm, index]),
+            }
+}
+CxPolicy [ result ] {
+
+	resourceMetricFilter := input.document[i].Resources
+    some name
+    resourceMetricFilter[name].Type == "AWS::Logs::MetricFilter"
+    
+    resourceAlarm := input.document[i].Resources
+    some nameAlarm
+    resourceAlarm[nameAlarm].Type == "AWS::CloudWatch::Alarm"
+    
+    
+	not regex.match("{ *\\$\\.userIdentity\\.type *= *Root *&& *\\$\\.userIdentity\\.invokedBy NOT EXISTS *&& *\\$\\.eventType *!= *AwsServiceEvent *}",
```

Comment:
Hello, I don't believe the regex will work with those following valid examples:

```
{ $.userIdentity.type = Root && $.userIdentity.invokedBy NOT EXISTS && $.eventType != AwsServiceEvent }
{ $.userIdentity.invokedBy NOT EXISTS && $.userIdentity.type = Root && $.eventType != AwsServiceEvent }
{ $.userIdentity.invokedBy NOT EXISTS && $.eventType != AwsServiceEvent && $.userIdentity.type = Root }
```

Reply:
In documentation it says 'Be sure you have typed the filter exactly as shown, and then click Assign Metric.'
This is the right filter: ' { $.userIdentity.type = "Root" && $.userIdentity.invokedBy NOT EXISTS && $.eventType != "AwsServiceEvent" } '
https://aws.amazon.com/blogs/security/how-to-receive-notifications-when-your-aws-accounts-root-access-keys-are-used/

Reply:
I understand, what I'm trying to say is that the conditions inside this filter could be permuted and should still be considered a valid case for the query.

Reply:
One way to check the filter regardless of the conditions order:
```
    filterPattern := trim(trim_space(resourceMetricFilter[name].Properties.FilterPattern), "{}")
    filterRulesArr := split(filterPattern, "&&")
    count(filterRulesArr) > 0
    
    condition1 := filterRulesArr[j]
    not regex.match(`\s*\$\.userIdentity\.type\s+=\s+Root\s*`, condition1)
    condition2 := filterRulesArr[k]
    not regex.match(`\s*\$\.userIdentity\.invokedBy\s+NOT\s+EXISTS\s*`, condition2)
    condition3 := filterRulesArr[l]
    not regex.match(`\s*\$\.eventType\s+!=\s+AwsServiceEvent\s*`, condition3)
```

Reply:
But this is yielding more results than expected. Just fruit for thought. I'll get back to this later.